### ISSUE TITLE
Can't resolve namedtuple parameters

### ISSUE THREAD

**Issue by spindensity** [microsoft/pyright#212]
Minimal reproducible case:

```python
import collections

Person = collections.namedtuple("Person", ["name", "age", "gender"])
bily = Person(name="Bily", age=28, gender="male")
```

![namedtuple](https://user-images.githubusercontent.com/33576822/63394669-4c889700-c3f3-11e9-8544-77a562981078.png)

Pyright: 1.0.56
VS Code: 1.37.1

**Comment by erictraut** [comment]
I'm not able to repro the problem as shown. Are you using any custom (non-default) configuration settings? Do you have a pyrightconfig.json file?

Is it possible there's a name collision with the "collections" module? In other words, do you have another "collections" package installed that would override the standard built-in typeshed files? If you right-click on "namedtuple" and choose "Go To Definition", where does it take you? For me, it takes me to typeshed-fallback/stdlib/3/collections/__init__.pyi.

**Comment by spindensity** [comment]
I don't have a pyrightconfig.json file.

When right-clicking on "namedtuple" and choose "Go To Definition",  vs code poped a window with two items to choose:

1. `typeshed-fallback/stdlib/3/collections/__init__.pyi`;
2. `lib\collections\__init__.py` in the root directory of the working environment.

![def1](https://user-images.githubusercontent.com/33576822/63397600-a726f080-c3fd-11e9-9d0c-15a9946f5c6d.png)
![def2](https://user-images.githubusercontent.com/33576822/63397605-abeba480-c3fd-11e9-95db-79067a26eaf5.png)

**Comment by erictraut** [comment]
I'm guessing that you have another VS Code extension installed that's also providing language servers for Python because Pyright is providing only one of these two definitions. It's probably the language service provided with the standard Microsoft Python extension. That shouldn't be a problem, but it is complicating the diagnosis.

Pyright will always use the typeshed definition in the ".pyi" file before any local ".py" files, unless it's configured otherwise, so it should be resolving the import with the `...collections/__init__.pyi`.

Please do the following:
1. Temporarily disable the standard Python extension (if my assumption above is correct). I want to eliminate any confusion that comes from having two language services running at the same time. You may need to reload the workspace after disabling it.
2. In the `import collections` statement, hover over the word "collections" and record a screen shot of the tool tip you see. I would expect it to look something like this:
![Screen Shot 2019-08-20 at 11 59 31 PM](https://user-images.githubusercontent.com/7040122/63409736-9f882d00-c3a6-11e9-9f7c-809c320757f0.png)
3. Enable the Python extension again.

**Comment by spindensity** [comment]
OK.

1. Disable standard python extension and reload workspace:
![proc1](https://user-images.githubusercontent.com/33576822/63417336-a7da7980-c433-11e9-96b8-d1b83129b836.png)

2. Hover over the word `collections` and `namedtuple`:
![proc2](https://user-images.githubusercontent.com/33576822/63417490-f8ea6d80-c433-11e9-90cb-f9bfc5cbd6bb.png)
![proc2 1](https://user-images.githubusercontent.com/33576822/63417501-fd168b00-c433-11e9-9039-0505f4f8a552.png)

3. Right click namedtuple and choose "Go to Definition", there's only one candidate so VS Code jumped to the right one directly:
![proc3](https://user-images.githubusercontent.com/33576822/63417808-96de3800-c434-11e9-945c-6ad9878cb916.png)

4. The error is still here:
![proc4](https://user-images.githubusercontent.com/33576822/63417921-c725d680-c434-11e9-93e1-a0a75a2e3c5f.png)

**Comment by spindensity** [comment]
Some more info, hope this helps:

The definition of `collections.namedtuple` in `typeshed-fallback/stdlib/3/collections/__init__.pyi` is:

```python
if sys.version_info >= (3, 7):
    def namedtuple(typename: str, field_names: Union[str, Iterable[str]], *,
                   rename: bool = ..., module: Optional[str] = ..., defaults: Optional[Iterable[Any]] = ...) -> Type[tuple]: ...
elif sys.version_info >= (3, 6):
    def namedtuple(typename: str, field_names: Union[str, Iterable[str]], *,
                   verbose: bool = ..., rename: bool = ..., module: Optional[str] = ...) -> Type[tuple]: ...
else:
    def namedtuple(typename: str, field_names: Union[str, Iterable[str]],
                   verbose: bool = ..., rename: bool = ...) -> Type[tuple]: ..
```

So the type of `Person` is `Type[tuple[Any]]` and all information of namedtuple parameters are erased.

I didi some debugging. When errors reported, the call stack of Pyright Language Server was:
```
_validateFunctionArguments (c:\Dev\Projects\pyright\server\src\analyzer\expressionEvaluator.ts:1973)
_validateCallArguments (c:\Dev\Projects\pyright\server\src\analyzer\expressionEvaluator.ts:1750)
_validateConstructorArguments (c:\Dev\Projects\pyright\server\src\analyzer\expressionEvaluator.ts:1691)
_getTypeFromCallExpressionWithBaseType (c:\Dev\Projects\pyright\server\src\analyzer\expressionEvaluator.ts:1595)
_getTypeFromCallExpression (c:\Dev\Projects\pyright\server\src\analyzer\expressionEvaluator.ts:1374)
_getTypeFromExpression (c:\Dev\Projects\pyright\server\src\analyzer\expressionEvaluator.ts:616)
getType (c:\Dev\Projects\pyright\server\src\analyzer\expressionEvaluator.ts:189)
_getTypeOfExpression (c:\Dev\Projects\pyright\server\src\analyzer\typeAnalyzer.ts:2894)
visitAssignment (c:\Dev\Projects\pyright\server\src\analyzer\typeAnalyzer.ts:1065)
visitNode (c:\Dev\Projects\pyright\server\src\analyzer\parseTreeWalker.ts:58)
walk (c:\Dev\Projects\pyright\server\src\analyzer\parseTreeWalker.ts:30)
walkMultiple.nodes.forEach.node (c:\Dev\Projects\pyright\server\src\analyzer\parseTreeWalker.ts:37)
walkMultiple (c:\Dev\Projects\pyright\server\src\analyzer\parseTreeWalker.ts:36)
walkChildren (c:\Dev\Projects\pyright\server\src\analyzer\parseTreeWalker.ts:42)
walk (c:\Dev\Projects\pyright\server\src\analyzer\parseTreeWalker.ts:31)
walkMultiple.nodes.forEach.node (c:\Dev\Projects\pyright\server\src\analyzer\parseTreeWalker.ts:37)
walkMultiple (c:\Dev\Projects\pyright\server\src\analyzer\parseTreeWalker.ts:36)
walkChildren (c:\Dev\Projects\pyright\server\src\analyzer\parseTreeWalker.ts:42)
walk (c:\Dev\Projects\pyright\server\src\analyzer\parseTreeWalker.ts:31)
analyze (c:\Dev\Projects\pyright\server\src\analyzer\typeAnalyzer.ts:111)
```

In the function `_getTypeFromCallExpressionWithBaseType` in `expressionEvaluator.ts`:

https://github.com/microsoft/pyright/blob/067c1e75d9bbba683ec8e46c8ee82d932a8da10b/server/src/analyzer/expressionEvaluator.ts#L1591-L1607
The `name` field value of `classFromTypeObject`  was`tuple` which is obviously right:
```
Object {name: "tuple", flags: 1, typeSourceId: 6658, …}
```
The length of argList was 3 and exactly corresponded to the arguments.

Nothing weird happened.

Then the expression `Person(name="Bily", age=28, gender="male")` was treated as a call to the constructor of tuple and `_validateConstructorArguments` was invoked.

The constructor of tuple does not have any named parameters, so errors were reported finally.

**Comment by erictraut** [comment]
Pyright (and other type checkers like Mypy) have special knowledge of the namedtuple function in the collections module. Pyright auto-generates the constructor methods for a namedtuple based on the parameters that are provided in the second (list) parameter. It also auto-generates instance variables and several other methods. It doesn't apply any of this special handling if the namedtuple function comes from a source other than the built-in collections module, so that's why I was interested to learn whether it was somehow picking up a different module called "collections". It looks like that's not the case, so we still have a mystery on our hands.

As I said, this all works correctly for me. I've copied your code directly and verified that it works. I also have several pyright unit tests that validate this particular case, and they're all working fine.

Thanks for diving in and doing some debugging. That's really helpful. If you don't mind doing some additional debugging, I recommend that you set a breakpoint on `_createNamedTupleType` in expressionEvaluator.ts. That provides the special-case handling that I mention above.

**Comment by spindensity** [comment]
After some further debugging, I can confirm it's a bug on windows.

Here is the detail:

`_createNamedTupleType` is never invoked. In the function `_getTypeFromCallExpressionWithBaseType`:
https://github.com/microsoft/pyright/blob/067c1e75d9bbba683ec8e46c8ee82d932a8da10b/server/src/analyzer/expressionEvaluator.ts#L1544-L1551
The conditon in line 1544 is always false when visiting node `collections.namedtuple` because the return value of `callType.getBuiltInName()` is always `undefined`.

Then we know, in the function `visitFunction`:
https://github.com/microsoft/pyright/blob/067c1e75d9bbba683ec8e46c8ee82d932a8da10b/server/src/analyzer/typeAnalyzer.ts#L297-L311
The two conditions in line 306 never become true.

Finally, in the file `sourceFile.ts`:
https://github.com/microsoft/pyright/blob/067c1e75d9bbba683ec8e46c8ee82d932a8da10b/server/src/analyzer/sourceFile.ts#L159-L172
`this._isTypingStubFile` is always false because the type stub file name is `__init__.pyi`, and `this._isBuiltInStubFile` is always false too since the following condition would never be true on **Windows** due to the `\` path seperator:
```python
this._filePath.endsWith('/collections/__init__.pyi')
```

The fix should be easy.

**Comment by erictraut** [comment]
Thanks, I really appreciate the debugging help! Yes, that makes sense. And yes, that's an easy bug to fix.

**Comment by erictraut** [comment]
I've checked in a fix for this bug (https://github.com/microsoft/pyright/commit/fef6830f1118be45198d6f068fa003dae3a862fd). It will be included in the next published version of pyright. Thanks again for your help!

**Comment by erictraut** [comment]
This is implemented in 1.0.57, which I just published. Thanks again for the bug report and the help with debugging.
